libjwt/jwt.c

```diff
@@ -663,6 +687,156 @@ int jwt_del_grant(jwt_t *jwt, const char *grant)
 
 #endif /* _MSC_VER */
 
+const char *jwt_get_header(jwt_t *jwt, const char *header)
+{
+	if (!jwt || !header || !strlen(header)) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	errno = 0;
+
+	return get_js_string(jwt->headers, header);
+}
+
+long jwt_get_header_int(jwt_t *jwt, const char *header)
+{
+	if (!jwt || !header || !strlen(header)) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	errno = 0;
+
+	return get_js_int(jwt->headers, header);
+}
+
+int jwt_get_header_bool(jwt_t *jwt, const char *header)
+{
+	if (!jwt || !header || !strlen(header)) {
+		errno = EINVAL;
+		return 0;
+	}
+
+	errno = 0;
+
+	return get_js_bool(jwt->headers, header);
+}
+
+char *jwt_get_headers_json(jwt_t *jwt, const char *header)
+{
+	json_t *js_val = NULL;
+
+	errno = EINVAL;
+
+	if (!jwt)
+		return NULL;
+
+	if (header && strlen(header))
+		js_val = json_object_get(jwt->headers, header);
+	else
+		js_val = jwt->headers;
+
+	if (js_val == NULL)
+		return NULL;
+
+	errno = 0;
+
+	return json_dumps(js_val, JSON_SORT_KEYS | JSON_COMPACT | JSON_ENCODE_ANY);
+}
+
+int jwt_add_header(jwt_t *jwt, const char *header, const char *val)
+{
+	if (!jwt || !header || !strlen(header) || !val)
+		return EINVAL;
+
+	if (get_js_string(jwt->headers, header) != NULL)
+		return EEXIST;
+
+	if (json_object_set_new(jwt->headers, header, json_string(val)))
+		return EINVAL;
+
+	return 0;
+}
+
+int jwt_add_header_int(jwt_t *jwt, const char *header, long val)
+{
+	if (!jwt || !header || !strlen(header))
+		return EINVAL;
+
+	if (get_js_int(jwt->headers, header) != -1)
+		return EEXIST;
+
+	if (json_object_set_new(jwt->headers, header, json_integer((json_int_t)val)))
+		return EINVAL;
+
+	return 0;
+}
+
+int jwt_add_header_bool(jwt_t *jwt, const char *header, int val)
+{
+	if (!jwt || !header || !strlen(header))
+		return EINVAL;
+
+	if (get_js_int(jwt->headers, header) != -1)
+		return EEXIST;
+
+	if (json_object_set_new(jwt->headers, header, json_boolean(val)))
+		return EINVAL;
+
+	return 0;
+}
+
+int jwt_add_headers_json(jwt_t *jwt, const char *json)
+{
+	json_t *js_val;
+	int ret = -1;
+
+	if (!jwt)
+		return EINVAL;
+
+	js_val = json_loads(json, JSON_REJECT_DUPLICATES, NULL);
+
+	if (json_is_object(js_val))
+		ret = json_object_update(jwt->headers, js_val);
+
+	json_decref(js_val);
+
+	return ret ? EINVAL : 0;
+}
+
+int jwt_del_headers(jwt_t *jwt, const char *header)
+{
+	if (!jwt)
+		return EINVAL;
+
+	if (header == NULL || !strlen(header))
+		json_object_clear(jwt->headers);
+	else
+		json_object_del(jwt->headers, header);
+
+	return 0;
+}
+
+#ifdef _MSC_VER
+
+int jwt_del_header(jwt_t *jwt, const char *header);
+#pragma comment(linker, "/alternatename:jwt_del_header=jwt_del_headers")
+
+#else
+
+#ifdef NO_WEAK_ALIASES
```

Comment:
`jwt_del_header` is unneeded. The only reason there is a counter for the body is that it is left as a deprecated function to avoid having to bump the API/ABI for the library. Let's not add a deprecated function. It can go.

Reply:
Sure.

Reply:
What about changing jwt_del_headers to jwt_del_header?

Reply:
I would say keep it the same name so it matches the semantics of the one for the grants.

Reply:
ok